### ISSUE TITLE
linter: typescript/consistent-return false positive on union

### ISSUE THREAD

**Issue by jz-cp** [oxc-project/oxc#22153]
### What version of Oxlint are you using?

1.63.0

### What command did you run?

oxlint

### What does your `.oxlintrc.json` (or `oxlint.config.ts`) config file look like?

```jsonc
// Replace this with your actual config file contents.
// It can be helpful to keep it minimal, if possible, but please ensure that it actually reproduces the issue you're reporting.
{
  "categories": {
    "pedantic": "deny",
    "suspicious": "deny",
  },
  "plugins": ["typescript"],
}
```

### What happened?

The code has:

1. An interface `Result` with a `status` property that is a union type of 3 possible strings: 'success' | 'failed' | 'loading'
2. An array of objects adhering to `Result` interface
3. Array.map() iterating over the array
4. Callback with a switch statement based on the status

All 3 items in the union are covered in the switch cases. I do not want to include a `default:` case which could make future maintenance difficult by masking additional statuses.

Playground link: (not this does not have the `typescript/consistent-return` enabled since it couldn't be added)
https://playground.oxc.rs/?options=%7B%22run%22%3A%7B%22lint%22%3Atrue%2C%22formatter%22%3Afalse%2C%22transform%22%3Afalse%2C%22isolatedDeclarations%22%3Afalse%2C%22whitespace%22%3Afalse%2C%22mangle%22%3Atrue%2C%22compress%22%3Afalse%2C%22scope%22%3Atrue%2C%22symbol%22%3Atrue%2C%22cfg%22%3Afalse%7D%2C%22parser%22%3A%7B%22extension%22%3A%22tsx%22%2C%22allowReturnOutsideFunction%22%3Atrue%2C%22preserveParens%22%3Atrue%2C%22allowV8Intrinsics%22%3Atrue%2C%22semanticErrors%22%3Atrue%7D%2C%22linter%22%3A%7B%22config%22%3A%22%7B%5C%22categories%5C%22%3A%7B%5C%22correctness%5C%22%3A%5C%22off%5C%22%7D%2C%5C%22rules%5C%22%3A%7B%5C%22eslint%2Farray-callback-return%5C%22%3A%5C%22error%5C%22%2C%5C%22eslint%2Fdefault-case%5C%22%3A%5C%22error%5C%22%7D%7D%22%7D%2C%22formatter%22%3A%7B%22useTabs%22%3Afalse%2C%22tabWidth%22%3A2%2C%22endOfLine%22%3A%22lf%22%2C%22printWidth%22%3A80%2C%22singleQuote%22%3Afalse%2C%22jsxSingleQuote%22%3Afalse%2C%22quoteProps%22%3A%22as-needed%22%2C%22trailingComma%22%3A%22all%22%2C%22semi%22%3Atrue%2C%22arrowParens%22%3A%22always%22%2C%22bracketSpacing%22%3Atrue%2C%22bracketSameLine%22%3Afalse%2C%22objectWrap%22%3A%22preserve%22%2C%22singleAttributePerLine%22%3Afalse%7D%2C%22transformer%22%3A%7B%22target%22%3A%22es2015%22%2C%22useDefineForClassFields%22%3Atrue%2C%22experimentalDecorators%22%3Atrue%2C%22emitDecoratorMetadata%22%3Atrue%2C%22optimizeEnums%22%3Atrue%2C%22optimizeConstEnums%22%3Atrue%7D%2C%22isolatedDeclarations%22%3A%7B%22stripInternal%22%3Afalse%7D%2C%22codegen%22%3A%7B%22normal%22%3Atrue%2C%22jsdoc%22%3Atrue%2C%22annotation%22%3Atrue%2C%22legal%22%3Atrue%7D%2C%22compress%22%3A%7B%7D%2C%22mangle%22%3A%7B%22topLevel%22%3Atrue%2C%22keepNames%22%3Afalse%7D%2C%22controlFlow%22%3A%7B%22verbose%22%3Afalse%7D%2C%22inject%22%3A%7B%22inject%22%3A%7B%7D%7D%2C%22define%22%3A%7B%22define%22%3A%7B%7D%7D%7D&code=interface+Result+%7B%0A++status%3A+%27success%27+%7C+%27failed%27+%7C+%27loading%27%3B%0A%7D%0A%0Aconst+results%3A+Result%5B%5D+%3D+%5B%0A++%7B+status%3A+%27success%27%7D%0A%5D%3B%0A%0Aresults.map%28result+%3D%3E+%7B%0A++switch+%28result.status%29+%7B%0A++++case+%27failed%27%3A%0A++++++return+%27...%27%3B%0A++++case+%27success%27%3A%0A++++++return+%27...%27%3B%0A++++case+%27loading%27%3A%0A++++++throw+new+Error%28%27...%27%29%0A++%7D%0A%7D%29%0A

Actual results:
Got an error:
```Function expected a return value. oxctypescript-eslint(consistent-return)```

Expected results:
No errors since all members of the union were accounted for, and this is a false positive.

Things I tried:

- Looking though the documentation for options to tune the behavior

**Comment by camc314** [comment]
Hi, I don't really understand what the problem is. It feels like a lot of issues are being mixed together here.

The `array-callback-return` doesn't use type information, and relies purely on control flow analysis.

Can you make isolated issues for each of the other rules that you believe are false positives?

**Comment by jz-cp** [comment]
@camc314, thanks for reading through this. Yes, there are two related issues here, and I can separate them. I'll remove the `array-callback-return` since it doesn't use type information, and update and simplify the title and description. However, the playground doesn't allow to specify typescript rules, even when the `typescript` plugin was specified. So I cannot provide a link replicating just the `typescript/consistent-return` issue.

**Comment by jz-cp** [comment]
Updated the title and desc

**Comment by camc314** [comment]
https://typescript-eslint.io/play/#ts=5.9.2&fileType=.tsx&code=JYOwLgpgTgZghgYwgAgEoQM4FcA2ZkDeAUMqchmHGFhgFzIDk2CSGGDyAPo-MDhABMO3BjgD2cAaADmDANxEAvkSIIxICsiiZcYOmh14A2gF1kAXmRGSZAuUrV9TLC0ztkykwqLbseDAB0ALZwAA4AFL66FgB8hDakGADuwGAIABbIkYZgARRUNACU8WSlyAhwGCgMvPxCtAllpdrUUCCMAZ3yjaUVVYzMrOwNTU0tWG0dXQqj5ZXV4pIyDCOzZGDpUGJJyCAQOwCiUFtQ4QydAQyFjcqK10A&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6MgeyeUuX0Ra1oifLGhN0URNGgdokcGAC%2BIJUA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

typescript-eslint reports the same behaviour as us.

I think you probably just don't want these lint rules enabled?

**Comment by RioShopfront** [comment]
typescript-eslint recommends that you use https://www.typescriptlang.org/tsconfig/#noImplicitReturns instead rather than the rule.

Also something like this addresses your concern about maintaining the default branch. It will produce a type error when modifications are made. https://typescript-eslint.io/play/#ts=5.9.2&fileType=.ts&code=JYOwLgpgTgZghgYwgAgEoQM4FcA2ZkDeAUMqchmHGFhgFzIDk2CSGGDyAPo-MDhABMO3BjgD2cAaADmwxgAcIIKSFkBuIgF8iRBGJAVkUTLjB00JvAG0AusgC8yKyTIFylauaZYWmdsk0AGhdSNwoqGnoGXn4hUm0bDSIIAA95MSh8PQN8ODZoMAA5CAA3aAdkAAoQUuh6GrKoAEp62qgHAD5CFzAACygxAHdkGuGAUSgBqEqAAwBVGrSIBEgBZGBIAFt6ABICBuhNGaaNTSTjbDwMADpNuHlKi9NO7rJyQY2EXqqnvGvwzxNV5vMgIPIoaJwPiCBi0EIgt7GahQECMa7ohgaBGkMEYCHMVjsOHYkFIrAotEYrHY3EQ8SSGSw%2BEkvoDYajZATKaVBjo64MJrMsgCCDwUzEklkMkUvJ4zLFRqPSxgf4eGiCt7aTRNIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6MgeyeUuX0Ra1oifLGhN0URNGgdokcGAC%2BIJUA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3QAacDUhFYASSSomyPACV0eNplIUwo9FIC%2BIQ0A&tokens=false   

As far as I can tell there isn't a false positive with consistent-returns
Code such as below should produce an error for noImplicit or consistent returns as the switch-case can fall past the "default" branch and just resolve/end, returning nothing.

```
results.map(result => {
    switch (result.status) {
        case 'failed':
            return '...';
        case 'success':
            return '...';
        case 'loading':
            throw new Error('...')
    }
})
```

**Comment by jz-cp** [comment]
@RioShopfront I thought that type information would/should be used in this analysis. And since all union members have been accounted for, there should be no need for a default. Further, I think putting a default there to satisfy the linter can lead to erroneous behavior if the union were ever modified.

**Comment by RioShopfront** [comment]
@jz-cp I think there is a balance to strike. In Typescript we can confirm that all branches are covered, but _technically_ the absent default branch could return implicitly in JS. Both the rule `typescript/consistent-return` and the `noImplicitReturns` from tsconfig raise an issue consistently here.

By adding the default branch in explicitly, rather than allowing the switch statement to just complete without hitting a branch, you are swapping hitting an error unknowingly, to another knowingly. This known error also has a guarantee that your own TS code won't hit it.

**Comment by RioShopfront** [comment]
https://github.com/typescript-eslint/typescript-eslint/issues/8577   A similar request that `consistent-returns` unnecessarily raises a warning over code which would satisfy something like https://typescript-eslint.io/rules/switch-exhaustiveness-check/     was made and then closed in favour of other analysis options.

> As far as I can tell there isn't a false positive with consistent-returns

I was partly wrong before, consistent-returns is behaving as expected in isolation, but combining it with types or other rules makes it redundant and/or wrong.

**Comment by camc314** [comment]
Thanks both for the additional color.


I agree that an exhaustive union switch is a useful pattern. To check for this, please use [typescript/switch-exhaustiveness-check](https://oxc.rs/docs/guide/usage/linter/rules/typescript/switch-exhaustiveness-check.html), as it is designed for this purpose.

The [typescript/consistent-return](https://oxc.rs/docs/guide/usage/linter/rules/typescript/consistent-return.html#typescript-consistent-return-oxlint) rule, checks if every runtime control-flow path consistently returns a value. A `switch` without a `default` can still fall through at runtime, so reporting this matches `@typescript-eslint/consistent-return` behavior. The equivalent request was also closed upstream: https://github.com/typescript-eslint/typescript-eslint/issues/8577.

If you want TypeScript’s control-flow analysis for missing returns, prefer `noImplicitReturns`, which both typescript-eslint and Oxlint’s docs recommend over this lint rule.


Closing this one out as not planned as this matches the expected behaviour of these two rules, and there are alternatives (mentioned above).


Thank both!

Please continue to report any other issues you have!
